tests/integration/version_migrations_test.go

```diff
@@ -0,0 +1,280 @@
+				},
+			},
+			env: []string{
+				fmt.Sprintf("POSTGRES_DB=%s", database),
+				fmt.Sprintf("POSTGRES_USER=%s", username),
+				fmt.Sprintf("POSTGRES_PASSWORD=%s", password),
+			},
+			cmd: []string{
+				"postgres",
+				"-cwal_level=logical",
+				"-cmax_replication_slots=1",
+				"-cmax_wal_senders=1",
+			},
+		})
+	if err != nil {
+		return "", 0, err
+	}
+
+	t.Cleanup(func() {
+		if err := docker.removeContainer(ctx, container.ID); err != nil {
+			t.Errorf("Could not remove container %s: %w", container.ID, err)
+		}
+	})
+
+	return container.ID, hostPort, nil
+}
+
+func testRow() *TestData {
+	row := TestData{}
+	row.text = "a test string"
+	row.boolean = false
+	row.date = time.Now()
+	row.json = `{"name": "Alice", "age": 31, "city": "LA"}`
+	row.jsonb = `{"name": "Bob", "age": 39, "city": "London"}`
+	row.array = []int32{1, 2, 3, 4, 5}
+	return &row
+}
+
+func writeTestData(config pgx.ConnConfig, n int, results chan<- int) {
+	rowsWritten := 0
+	conn, err := pgx.Connect(config)
+	if err != nil {
+		results <- rowsWritten
+		return
+	}
+	defer conn.Close()
+	for i := 0; i < n; i++ {
+		row := testRow()
+		_, err = conn.Exec(insertSQL, row.text, row.date, row.boolean, row.json, row.jsonb, row.array)
+		if err != nil {
+			fmt.Printf("oops: %v", err)
+			results <- rowsWritten
+			return
+		}
+		rowsWritten = rowsWritten + 1
+	}
+	results <- rowsWritten
+}
+
+func TestVersionMigration(t *testing.T) {
+	testCases := []struct {
+		name   string
+		source string
+		target string
+	}{
+		{
+			name:   "9.5To9.6",
+			source: "9.5",
+			target: "9.6",
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			ctx := context.Background()
+
+			//bring up source and target database containers
+			_, srcPort, err := createDatabaseContainer(t, ctx, tc.source, dbUser, dbPassword, dbName)
+			require.NoError(t, err)
+			srcDBConfig := pgx.ConnConfig{
+				Host:     "127.0.0.1",
+				Port:     uint16(srcPort),
+				User:     dbUser,
+				Password: dbPassword,
+				Database: dbName,
+			}
+			err = setupTestSchema(srcDBConfig)
+			require.NoError(t, err)
+
+			_, targetPort, err := createDatabaseContainer(t, ctx, tc.target, dbUser, dbPassword, dbName)
+			require.NoError(t, err)
+			targetDBConfig := pgx.ConnConfig{
+				Host:     "127.0.0.1",
+				Port:     uint16(targetPort),
+				User:     dbUser,
+				Password: dbPassword,
+				Database: dbName,
+			}
+			err = setupTestSchema(targetDBConfig)
+			require.NoError(t, err)
+
+			// setup warp-pipe on source database
+			wpConn, err := pgx.Connect(srcDBConfig)
+			require.NoError(t, err)
+			err = db.Prepare(wpConn, []string{"public"}, []string{"test"}, []string{})
+			if err != nil {
+				t.Errorf("Could not setup warp pipe: %v", err)
+			}
+
+			// write rows to the source database
+			writersCount := 20
+			rowsWritten := make(chan int, writersCount)
+			for i := 0; i < writersCount; i++ {
+				go writeTestData(srcDBConfig, 50, rowsWritten)
```

Comment:
Why did this need to be a go routine? And why the channel for rowsWritten? I would have thought the whole process could be serialized? Axon will block until the sync is complete.

Reply:
I thought this would be closer to what actually happens during a migration. The source DB will still accept writes while warppipe is setup and Axon is running. At the time of writing, I used the rowsWritten to check that some part of the sync work as expected: get the number of rows written to the source database and then ensure the same number of rows were written to the target database. This can be replaced with the checksum approach in yoshi.